fili-core/src/main/java/com/yahoo/bard/webservice/table/availability/MetricUnionAvailability.java

```diff
@@ -0,0 +1,125 @@
+package com.yahoo.bard.webservice.table.availability;
+
+import com.yahoo.bard.webservice.data.config.names.TableName;
+import com.yahoo.bard.webservice.table.Column;
+import com.yahoo.bard.webservice.table.PhysicalTable;
+import com.yahoo.bard.webservice.table.Table;
+import com.yahoo.bard.webservice.table.resolver.DataSourceConstraint;
+import com.yahoo.bard.webservice.util.IntervalUtils;
+import com.yahoo.bard.webservice.util.SimplifiedIntervalList;
+
+import org.joda.time.Interval;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import javax.validation.constraints.NotNull;
+
+/**
+ * An availability which unions multiple different physical tables together into a single table
+ * that contains all columns from all tables so that we can query different columns
+ * from different tables at the same time.
+ */
+public class MetricUnionAvailability implements Availability {
+
+    Set<Column> columns;
+    Set<TableName> dataSourceNames;
+    Map<Availability, Set<String>> metricAvailability;
+
+    /**
+     * Constructor.
+     *
+     * @param metricAvailability A mapping from a table to its contents availability
+     * @param columns The columns associated with all tables
+     */
+    public MetricUnionAvailability(
+            @NotNull Map<Table, Set<String>> metricAvailability,
+            @NotNull Set<Column> columns
+    ) {
+        // validate metric uniqueness such that
+        // each table's underlying datasource schema do not have repeated metric column
+        if (metricAvailability.values().stream()
+                .flatMap(Set::stream)
+                .collect(
+                        Collectors.groupingBy(
+                                Function.identity(),
+                                Collectors.counting()
+                        )
+                )
+                .entrySet()
+                .stream()
+                .anyMatch(entry -> entry.getValue() > 1)) {
+            throw new IllegalArgumentException("Metric columns are not unique.");
+        }
+
+        this.columns = columns;
+        this.dataSourceNames = metricAvailability.keySet().stream()
+                .map(table -> TableName.of(table.getName()))
+                .collect(Collectors.toSet());
+        this.metricAvailability = metricAvailability.entrySet().stream()
+                .collect(
+                        Collectors.toMap(
+                                entry -> PhysicalTable.class.cast(entry.getKey()).getAvailability(),
+                                Map.Entry::getValue
+                        )
+                );
+    }
+
+    @Override
+    public Set<TableName> getDataSourceNames() {
+        return dataSourceNames;
+    }
+
+    @Override
+    public Map<Column, Set<Interval>> getAllAvailableIntervals() {
+        return metricAvailability.entrySet().stream()
```

Comment:
I think there might and will be repeated columns, we need to union those availabilities together I think using group by.